pageserver/src/http/routes.rs

```diff
@@ -1091,6 +1175,64 @@ async fn post_tracing_event_handler(mut r: Request<Body>) -> Result<Response<Bod
     json_response(StatusCode::OK, ())
 }
 
+/// Common functionality of all the HTTP API handlers.
+///
+/// - Adds a tracing span to each request (by `request_span`)
+/// - Logs the request depending on the request method (by `request_span`)
+/// - Logs the response if it was not successful (by `request_span`
+/// - Shields the handler function from async cancellations. Hyper can drop the handler
+///   Future if the connection to the client is lost, but most of the pageserver code is
+///   not async cancellation safe. This converts the dropped future into a graceful cancellation
+///   request with a CancellationToken.
+async fn api_handler<R, H>(request: Request<Body>, handler: H) -> Result<Response<Body>, ApiError>
+where
+    R: std::future::Future<Output = Result<Response<Body>, ApiError>> + Send + 'static,
+    H: FnOnce(Request<Body>, CancellationToken) -> R + Send + Sync + 'static,
+{
+    // Spawn a new task to handle the request, to protect the handler from unexpected
+    // async cancellations. Most pageserver functions are not async cancellation safe.
+    // We arm a drop-guard, so that if Hyper drops the Future, we signal the task
+    // with the cancellation token.
+    let token = CancellationToken::new();
+    let guard = token.clone().drop_guard();
+    let handle = tokio::spawn(request_span(request, move |r| handler(r, token)));
+
+    let result = handle.await;
+
+    guard.disarm();
+
+    match result {
+        Ok(result) => result,
+        Err(e) => {
+            error!("HTTP request handler task panicked: {e:?}");
```

Comment:
Is this useful if the panic happened? We would had already logged it via panic handler, at worst this will output it as two errors.

We most likely don't need to log anything here.

You should be able to test this with `curl localhost:9898/v1/panic?message=hello`.

Reply:
You're right, we already log it in the panic handler. I modified this to print just a brief message to the log. I also changed it to return an `Ok` with an HTTP error response, because on `Err`, the generic `error::handler` logs the error and there's no need for it there either.